Panic in (*DatabaseContext).getChangesInChannelFromView

I just noticed that my local Sync Gateway instance that I use for Couchbase Lite unit tests had crashed. [The panic dump is here](https://gist.github.com/snej/fcf840f219b0e4409864), with some of the last log output at the top.

The panicked thread looks like:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0xc0 pc=0x220e03]

goroutine 13898 [running]:
github.com/couchbase/sync_gateway/db.(*DatabaseContext).getChangesInChannelFromView(0xc210fba540, 0x6e73f0, 0x1, 0x38d, 0x0, 0x0, 0x38d, 0x0, 0x0, 0x0, ...)
	/Couchbase/sync_gateway/src/github.com/couchbase/sync_gateway/db/changes_view.go:36 +0x3f3
github.com/couchbase/sync_gateway/db.func·005(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Couchbase/sync_gateway/src/github.com/couchbase/sync_gateway/db/change_cache.go:190 +0x226
github.com/couchbase/sync_gateway/db.(*changeCache).CleanSkippedSequenceQueue(0xc210fba610, 0x2501)
	/Couchbase/sync_gateway/src/github.com/couchbase/sync_gateway/db/change_cache.go:205 +0x63
github.com/couchbase/sync_gateway/db.func·004()
	/Couchbase/sync_gateway/src/github.com/couchbase/sync_gateway/db/change_cache.go:119 +0x34
created by github.com/couchbase/sync_gateway/db.(*changeCache).Init
	/Couchbase/sync_gateway/src/github.com/couchbase/sync_gateway/db/change_cache.go:122 +0x3c3
```

Also worth noting is that there are 182 goroutines with the following stack, which leads me to think that they're not getting cleaned up properly:
```
goroutine 13897 [sleep]:
github.com/couchbase/sync_gateway/db.func·003()
	/Couchbase/sync_gateway/src/github.com/couchbase/sync_gateway/db/change_cache.go:113 +0x66
created by github.com/couchbase/sync_gateway/db.(*changeCache).Init
	/Couchbase/sync_gateway/src/github.com/couchbase/sync_gateway/db/change_cache.go:115 +0x371
```

--- a/src/github.com/couchbase/sync_gateway/db/change_cache.go
+++ b/src/github.com/couchbase/sync_gateway/db/change_cache.go
@@ -109,15 +109,21 @@ func (c *changeCache) Init(context *DatabaseContext, lastSequence uint64, onChan
 
 	// Start a background task for periodic housekeeping:
 	go func() {
-		for c.CleanUp() {
+		for {
 			time.Sleep(c.options.CachePendingSeqMaxWait / 2)
+			if c.stopped || !c.CleanUp() {
+				break
+			}
 		}
 	}()
 
 	// Start a background task for SkippedSequenceQueue housekeeping:
 	go func() {
-		for c.CleanSkippedSequenceQueue() {
+		for {
 			time.Sleep(c.options.CacheSkippedSeqMaxWait / 2)
+			if c.stopped || !c.CleanSkippedSequenceQueue() {
+				break
+			}
 		}
 	}()
 }

--- a/src/github.com/couchbase/sync_gateway/db/change_cache.go
+++ b/src/github.com/couchbase/sync_gateway/db/change_cache.go
@@ -109,21 +109,17 @@ func (c *changeCache) Init(context *DatabaseContext, lastSequence uint64, onChan
 
 	// Start a background task for periodic housekeeping:
 	go func() {
-		for {
+		time.Sleep(c.options.CachePendingSeqMaxWait / 2)
+		for !c.IsStopped() && c.CleanUp() {
 			time.Sleep(c.options.CachePendingSeqMaxWait / 2)
-			if c.stopped || !c.CleanUp() {
-				break
-			}
 		}
 	}()
 
 	// Start a background task for SkippedSequenceQueue housekeeping:
 	go func() {
-		for {
+		time.Sleep(c.options.CacheSkippedSeqMaxWait / 2)
+		for !c.IsStopped() && c.CleanSkippedSequenceQueue() {
 			time.Sleep(c.options.CacheSkippedSeqMaxWait / 2)
-			if c.stopped || !c.CleanSkippedSequenceQueue() {
-				break
-			}
 		}
 	}()
 }
@@ -134,6 +130,12 @@ func (c *changeCache) Stop() {
 	c.stopped = true
 	c.logsDisabled = true
 	c.lock.Unlock()
+}
+
+func (c *changeCache) IsStopped() bool {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.stopped
 }
 
 // Forgets all cached changes for all channels.
@@ -560,7 +562,7 @@ func (c *changeCache) _getChannelCache(channelName string) *channelCache {
 //////// CHANGE ACCESS:
 
 func (c *changeCache) GetChangesInChannel(channelName string, options ChangesOptions) ([]*LogEntry, error) {
-	if c.stopped {
+	if c.IsStopped() {
 		return nil, base.HTTPErrorf(503, "Database closed")
 	}
 	return c.getChannelCache(channelName).GetChanges(options)

--- a/src/github.com/couchbase/sync_gateway/db/change_cache_test.go
+++ b/src/github.com/couchbase/sync_gateway/db/change_cache_test.go
@@ -730,6 +730,35 @@ func TestSkippedViewRetrieval(t *testing.T) {
 
 }
 
+// Test that housekeeping goroutines get terminated when change cache is stopped
+func TestStopChangeCache(t *testing.T) {
+	// Setup short-wait cache to ensure cleanup goroutines fire often
+	cacheOptions := CacheOptions{
+		CachePendingSeqMaxWait: 10 * time.Millisecond,
+		CachePendingSeqMaxNum:  50,
+		CacheSkippedSeqMaxWait: 1 * time.Second}
+	// Use leaky bucket to have the tap feed 'lose' document 3
+	leakyConfig := base.LeakyBucketConfig{
+		TapFeedMissingDocs: []string{"doc-3"},
+	}
+	db := setupTestLeakyDBWithCacheOptions(t, cacheOptions, leakyConfig)
+
+	// Write sequences direct
+	WriteDirect(db, []string{"ABC"}, 1)
+	WriteDirect(db, []string{"ABC"}, 2)
+	WriteDirect(db, []string{"ABC"}, 3)
+
+	// Artificially add 3 skipped, and back date skipped entry by 2 hours to trigger attempted view retrieval during Clean call
+	db.changeCache.skippedSeqs.Push(&SkippedSequence{3, time.Now().Add(time.Duration(time.Hour * -2))})
+
+	// tear down the DB.  Should stop the cache before view retrieval of the skipped sequence is attempted.
+	tearDownTestDB(t, db)
+
+	// Hang around a while to see if the housekeeping tasks fire and panic
+	time.Sleep(2 * time.Second)
+
+}
+
 func shortWaitCache() CacheOptions {
 
 	return CacheOptions{

--- a/src/github.com/couchbase/sync_gateway/db/changes_view.go
+++ b/src/github.com/couchbase/sync_gateway/db/changes_view.go
@@ -1,6 +1,7 @@
 package db
 
 import (
+	"errors"
 	"time"
 
 	"github.com/couchbase/go-couchbase"
@@ -28,6 +29,9 @@ type channelsViewRow struct {
 // Queries the 'channels' view to get a range of sequences of a single channel as LogEntries.
 func (dbc *DatabaseContext) getChangesInChannelFromView(
 	channelName string, endSeq uint64, options ChangesOptions) (LogEntries, error) {
+	if dbc.Bucket == nil {
+		return nil, errors.New("No bucket available for channel view query")
+	}
 	start := time.Now()
 	// Query the view:
 	optMap := changesViewOptions(channelName, endSeq, options)

